docs/user-tutorial/data-diagnosis.md

```diff
@@ -0,0 +1,144 @@
+        - ib-loopback/return_code
+    rule0:
+    # Rule 0: If KernelLaunch suffers > 5% downgrade, label it as defective
+      function: variance
+      criteria: '>,5%'
+      categories: KernelLaunch
+      metrics:
+        - kernel-launch/event_overhead:\d+
+        - kernel-launch/wall_overhead:\d+
+    rule1:
+    # Rule 1: If H2D_Mem_BW or D2H_Mem_BW test suffers > 5% downgrade, label it as defective
+      function: variance
+      criteria: '<,-5%'
+      categories: Mem
+      metrics:
+        - mem-bw/H2D_Mem_BW:\d+
+        - mem-bw/D2H_Mem_BW:\d+
+    rule2:
+    # Rule 2: If NCCL_BW suffers > 5% downgrade, label it as defective
+      function: variance
+      criteria: '<,-5%'
+      categories: NCCL
+      metircs:
+        - nccl-bw/allreduce_8589934592_busbw:0
+    rule3:
+    # Rule 3: If GPT-2, BERT suffers > 5% downgrade, label it as defective
+      function: variance
+      criteria: '<,-5%'
+      categories: Model
+      metrics:
+        - bert_models/pytorch-bert-base/throughput_train_float(32|16)
+        - bert_models/pytorch-bert-large/throughput_train_float(32|16)
+        - gpt_models/pytorch-gpt-large/throughput_train_float(32|16)
+```
+
+This Rule file describes the rules used for data diagnosis.
+
+They are firstly organized by the rule name, and each rule mainly includes 4 elements:
+
+#### `metrics`
+
+The list of metrics for this rule. Each metric is in the format of ${benchmark_name}/regex, you can use regex after the first '/', but to be noticed, the benchmark name can not be a regex.
+
+#### `categories`
+
+The categories belong to this rule.
+
+#### `criteria`
+
+The criteria used for this rule, which indicate how to compare the data with the baseline value. The format is `${operator},${value}`.
+
+- ${operator}: comparison operator, < or >
+- ${value}: the threshold, float or percentage
+
+#### `function`
+
+The function used for this rule.
+
+2 types of rules are supported currently:
+
+- `variance`: the rule is that if the variance between raw data and baseline violates the criteria. variance = (raw data - criteria) / criteria
+
+  For example, if the criteria are `>,5%`, the rule is that if the variance is larger than 5%, it should be defective.
+
+- `value`: the rule is that if the raw data violate the criteria.
+
+  For example, if the criteria are `>,0`, the rule is that if the raw data is larger than the 0, it should be defective.
+
+`Tips`: you must contain a default rule for ${benchmark_name}/return_code as the above in the example, which is used to identify failed tests.
+
+## Output
+
+We support different output formats for filtering the defective machines including jsonl, excel, etc. The output includes all defective servers' information including index, failure category, failure details, and detailed metrics.
+
+- index: the name of defective machines.
+
+- Category: categories defined in the rule.
+
+- Issue Details: all violated metrics including metric data and related rule.
```

Comment:
Can we use `Defective Details`? Looks we don't use the term `Issue` in the whole doc except for here.

Reply:
sure, updated